src/client/communicator.go

```diff
@@ -64,41 +64,51 @@ func (this *Communicator) GetVmsData() (interface{}, error) {
 	if err != nil {
 		return nil, fmt.Errorf("getting vms data error: %s", err)
 	}
-	s, err := this.GetRegistryVms()
+	s, err := this.GetRegistryTemplatesData()
 	if err != nil {
-		return nil, fmt.Errorf("getting vms registry data error: %s", err)
+		return nil, fmt.Errorf("getting registry templates data error: %s", err)
 	}
+	templates := s.([]types.Template)
 	m := d.([]types.Instance)
-	for k, v := range m {
-		i := v.Vm.TemplateUUID
-		m[k].Vm.TemplateNAME = s[i]
+	for i, v := range m {
+		for _, templateV := range templates {
+			if templateV.UUID == v.Vm.TemplateUUID {
+				m[i].Vm.TemplateNAME = templateV.Name
+			}
+		}
 	}
 	return m, nil
 }
 
-func (this *Communicator) GetRegistryData() (interface{}, error) {
+func (this *Communicator) GetRegistryDiskData() (interface{}, error) {
 	endpoint := "/api/v1/registry/disk"
-	resp := &types.RegistryResponse{}
+	resp := &types.RegistryDiskResponse{}
 	d, err := this.getData(endpoint, resp)
 	if err != nil {
-		return nil, fmt.Errorf("getting registry data error: %s", err)
+		return nil, fmt.Errorf("getting registry disk data error: %s", err)
 	}
 	return d, nil
 }
 
-func (this *Communicator) GetRegistryVms() (map[string]string, error) {
+func (this *Communicator) GetRegistryTemplatesData() (interface{}, error) {
 	endpoint := "/api/v1/registry/vm"
-	resp := &types.RegistryVmResponse{}
-	d, err := this.getData(endpoint, resp)
+	resp := &types.RegistryTemplateResponse{}
+	templates, err := this.getData(endpoint, resp)
 	if err != nil {
-		return nil, fmt.Errorf("getting registry vms error: %s", err)
-	}
-	m := d.([]types.Vms)
-	VmsData := make(map[string]string)
-	for _, v := range m {
-		VmsData[v.VmID] = v.TemplateNAME
-	}
-	return VmsData, nil
+		return nil, fmt.Errorf("getting registry templates error: %s", err)
+	}
+	templatesArray := templates.([]types.Template)
+	for i, v := range templatesArray {
+		endpoint := "/api/v1/registry/vm?id=" + v.UUID
```

Comment:
I am worried about the growing amount of reuqests required to update the VM data. This PR adds a request PER template, not so much scalable.
Fetching the VM data can happen every few seconds, while in fact the templates information updates once in a while.

what do you think about creating an in-memory template data that is updated once every minute (can be adjusted), and then when the VM data is fetched it uses that in-memory data instead of sending new reuests to the controller for information?

Reply:
I wonder if we could add a count of tags to the registry/vm endpoint and then only if that changes, make the query for the tag specifics. If the user quickly deletes a tag and then pushes another (keeping the tag count the same), that may not update Prometheus, but what are the odds the 15 seconds necessary to query for the new count value in the API would be longer than the time it takes for a user to delete and then push a new tag. Something to consider, but I don't know if we can solve this right now.

Reply:
This is a build and run without any tag changes:

```
❯ make build-and-run
kill -15 $(pgrep "[a]nka-prometheus") || true
kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]
/Applications/Xcode.app/Contents/Developer/usr/bin/make build-mac
GOOS=darwin /Applications/Xcode.app/Contents/Developer/usr/bin/make build RACE="-race"
GOARCH=amd64 go build -race -ldflags "-X main.version=2.2.2" -o bin/anka-prometheus-exporter_darwin_amd64
chmod +x bin/anka-prometheus-exporter_darwin_amd64
./bin/anka-prometheus-exporter_darwin_amd64 --controller-address http://anka.controller:8090
{"level":"info","msg":"Starting Prometheus Exporter for Anka (2.2.2)","time":"2021-06-17T15:58:50-04:00"}
TEMPLATE MAP: map[]
TEMPLATE SIZE: 0 32512360448
TEMPLATE SIZE: 0 29855444992
TEMPLATE SIZE: 0 21666447360
TEMPLATE SIZE: 0 26104016896
TEMPLATE SIZE: 0 21062053888
{"level":"info","msg":"Serving metrics at /metrics and :2112","time":"2021-06-17T15:58:50-04:00"}
TEMPLATE MAP: map[03160cb4-8509-48b6-9f82-fd009638e95f:{UUID:03160cb4-8509-48b6-9f82-fd009638e95f Name:11.4-openjdk-1.8.0_292-teamcity Size:32512360448 Tags:[{Name:v1 Size:32512360448}]} 5d1b40b9-7e68-4807-a290-c59c66e926b4:{UUID:5d1b40b9-7e68-4807-a290-c59c66e926b4 Name:11.4-gitlab Size:29855444992 Tags:[{Name:v1 Size:25307734016} {Name:v1-with-file Size:25868128256}]} a7487422-3c3c-4ab9-9d9a-e886af92c080:{UUID:a7487422-3c3c-4ab9-9d9a-e886af92c080 Name:11.4-openjdk-1.8.0_292 Size:21666447360 Tags:[{Name:v1 Size:21666447360}]} c0847bc9-5d2d-4dbc-ba6a-240f7ff08032:{UUID:c0847bc9-5d2d-4dbc-ba6a-240f7ff08032 Name:11.4-openjdk-1.8.0_292-jenkins Size:26104016896 Tags:[{Name:v1 Size:26104016896}]} c12ccfa5-8757-411e-9505-128190e9854e:{UUID:c12ccfa5-8757-411e-9505-128190e9854e Name:11.4 Size:21062053888 Tags:[{Name:vanilla Size:18572955648} {Name:vanilla+port-forward-22 Size:18572955648} {Name:vanilla+port-forward-22+brew-git Size:21062053888}]}]
2021/06/17 15:58:50 getVmsData took 17.739766ms
2021/06/17 15:59:05 getVmsData took 8.199642ms
TEMPLATE MAP: map[03160cb4-8509-48b6-9f82-fd009638e95f:{UUID:03160cb4-8509-48b6-9f82-fd009638e95f Name:11.4-openjdk-1.8.0_292-teamcity Size:32512360448 Tags:[{Name:v1 Size:32512360448}]} 5d1b40b9-7e68-4807-a290-c59c66e926b4:{UUID:5d1b40b9-7e68-4807-a290-c59c66e926b4 Name:11.4-gitlab Size:29855444992 Tags:[{Name:v1 Size:25307734016} {Name:v1-with-file Size:25868128256}]} a7487422-3c3c-4ab9-9d9a-e886af92c080:{UUID:a7487422-3c3c-4ab9-9d9a-e886af92c080 Name:11.4-openjdk-1.8.0_292 Size:21666447360 Tags:[{Name:v1 Size:21666447360}]} c0847bc9-5d2d-4dbc-ba6a-240f7ff08032:{UUID:c0847bc9-5d2d-4dbc-ba6a-240f7ff08032 Name:11.4-openjdk-1.8.0_292-jenkins Size:26104016896 Tags:[{Name:v1 Size:26104016896}]} c12ccfa5-8757-411e-9505-128190e9854e:{UUID:c12ccfa5-8757-411e-9505-128190e9854e Name:11.4 Size:21062053888 Tags:[{Name:vanilla Size:18572955648} {Name:vanilla+port-forward-22 Size:18572955648} {Name:vanilla+port-forward-22+brew-git Size:21062053888}]}]
```

You can see thing stay the same and TEMPLATE SIZE isn't output, so this is never triggered:

```
	templatesArray := templates.([]types.Template)
	templatesMap := state.GetState().GetTemplatesMap()
	fmt.Printf("TEMPLATE MAP: %+v\n", templatesMap)
	for i, template := range templatesArray {
		if templatesMap[template.UUID].Size != template.Size {
			fmt.Printf("TEMPLATE SIZE: %+v %+v\n", templatesMap[template.UUID].Size, template.Size)
			endpoint := "/api/v1/registry/vm?id=" + template.UUID
```

Then, let's run it again and push a new tag:

```
{"level":"info","msg":"Starting Prometheus Exporter for Anka (2.2.2)","time":"2021-06-17T15:59:20-04:00"}
TEMPLATE MAP: map[]
TEMPLATE SIZE: 0 32512360448
TEMPLATE SIZE: 0 29855444992
TEMPLATE SIZE: 0 21666447360
TEMPLATE SIZE: 0 26104016896
TEMPLATE SIZE: 0 21062053888
2021/06/17 15:59:20 getVmsData took 1.753719ms
{"level":"info","msg":"Serving metrics at /metrics and :2112","time":"2021-06-17T15:59:20-04:00"}
TEMPLATE MAP: map[03160cb4-8509-48b6-9f82-fd009638e95f:{UUID:03160cb4-8509-48b6-9f82-fd009638e95f Name:11.4-openjdk-1.8.0_292-teamcity Size:32512360448 Tags:[{Name:v1 Size:32512360448}]} 5d1b40b9-7e68-4807-a290-c59c66e926b4:{UUID:5d1b40b9-7e68-4807-a290-c59c66e926b4 Name:11.4-gitlab Size:29855444992 Tags:[{Name:v1 Size:25307734016} {Name:v1-with-file Size:25868128256}]} a7487422-3c3c-4ab9-9d9a-e886af92c080:{UUID:a7487422-3c3c-4ab9-9d9a-e886af92c080 Name:11.4-openjdk-1.8.0_292 Size:21666447360 Tags:[{Name:v1 Size:21666447360}]} c0847bc9-5d2d-4dbc-ba6a-240f7ff08032:{UUID:c0847bc9-5d2d-4dbc-ba6a-240f7ff08032 Name:11.4-openjdk-1.8.0_292-jenkins Size:26104016896 Tags:[{Name:v1 Size:26104016896}]} c12ccfa5-8757-411e-9505-128190e9854e:{UUID:c12ccfa5-8757-411e-9505-128190e9854e Name:11.4 Size:21062053888 Tags:[{Name:vanilla Size:18572955648} {Name:vanilla+port-forward-22 Size:18572955648} {Name:vanilla+port-forward-22+brew-git Size:21062053888}]}]




2021/06/17 15:59:35 getVmsData took 1.487986ms
TEMPLATE MAP: map[03160cb4-8509-48b6-9f82-fd009638e95f:{UUID:03160cb4-8509-48b6-9f82-fd009638e95f Name:11.4-openjdk-1.8.0_292-teamcity Size:32512360448 Tags:[{Name:v1 Size:32512360448}]} 5d1b40b9-7e68-4807-a290-c59c66e926b4:{UUID:5d1b40b9-7e68-4807-a290-c59c66e926b4 Name:11.4-gitlab Size:29855444992 Tags:[{Name:v1 Size:25307734016} {Name:v1-with-file Size:25868128256}]} a7487422-3c3c-4ab9-9d9a-e886af92c080:{UUID:a7487422-3c3c-4ab9-9d9a-e886af92c080 Name:11.4-openjdk-1.8.0_292 Size:21666447360 Tags:[{Name:v1 Size:21666447360}]} c0847bc9-5d2d-4dbc-ba6a-240f7ff08032:{UUID:c0847bc9-5d2d-4dbc-ba6a-240f7ff08032 Name:11.4-openjdk-1.8.0_292-jenkins Size:26104016896 Tags:[{Name:v1 Size:26104016896}]} c12ccfa5-8757-411e-9505-128190e9854e:{UUID:c12ccfa5-8757-411e-9505-128190e9854e Name:11.4 Size:21062053888 Tags:[{Name:vanilla Size:18572955648} {Name:vanilla+port-forward-22 Size:18572955648} {Name:vanilla+port-forward-22+brew-git Size:21062053888}]}]
2021/06/17 15:59:50 getVmsData took 1.168641ms
TEMPLATE MAP: map[03160cb4-8509-48b6-9f82-fd009638e95f:{UUID:03160cb4-8509-48b6-9f82-fd009638e95f Name:11.4-openjdk-1.8.0_292-teamcity Size:32512360448 Tags:[{Name:v1 Size:32512360448}]} 5d1b40b9-7e68-4807-a290-c59c66e926b4:{UUID:5d1b40b9-7e68-4807-a290-c59c66e926b4 Name:11.4-gitlab Size:29855444992 Tags:[{Name:v1 Size:25307734016} {Name:v1-with-file Size:25868128256}]} a7487422-3c3c-4ab9-9d9a-e886af92c080:{UUID:a7487422-3c3c-4ab9-9d9a-e886af92c080 Name:11.4-openjdk-1.8.0_292 Size:21666447360 Tags:[{Name:v1 Size:21666447360}]} c0847bc9-5d2d-4dbc-ba6a-240f7ff08032:{UUID:c0847bc9-5d2d-4dbc-ba6a-240f7ff08032 Name:11.4-openjdk-1.8.0_292-jenkins Size:26104016896 Tags:[{Name:v1 Size:26104016896}]} c12ccfa5-8757-411e-9505-128190e9854e:{UUID:c12ccfa5-8757-411e-9505-128190e9854e Name:11.4 Size:21062053888 Tags:[{Name:vanilla Size:18572955648} {Name:vanilla+port-forward-22 Size:18572955648} {Name:vanilla+port-forward-22+brew-git Size:21062053888}]}]
TEMPLATE SIZE: 21062053888 23465168896
2021/06/17 16:00:05 getVmsData took 1.75797ms
TEMPLATE MAP: map[03160cb4-8509-48b6-9f82-fd009638e95f:{UUID:03160cb4-8509-48b6-9f82-fd009638e95f Name:11.4-openjdk-1.8.0_292-teamcity Size:32512360448 Tags:[{Name:v1 Size:32512360448}]} 5d1b40b9-7e68-4807-a290-c59c66e926b4:{UUID:5d1b40b9-7e68-4807-a290-c59c66e926b4 Name:11.4-gitlab Size:29855444992 Tags:[{Name:v1 Size:25307734016} {Name:v1-with-file Size:25868128256}]} a7487422-3c3c-4ab9-9d9a-e886af92c080:{UUID:a7487422-3c3c-4ab9-9d9a-e886af92c080 Name:11.4-openjdk-1.8.0_292 Size:21666447360 Tags:[{Name:v1 Size:21666447360}]} c0847bc9-5d2d-4dbc-ba6a-240f7ff08032:{UUID:c0847bc9-5d2d-4dbc-ba6a-240f7ff08032 Name:11.4-openjdk-1.8.0_292-jenkins Size:26104016896 Tags:[{Name:v1 Size:26104016896}]} c12ccfa5-8757-411e-9505-128190e9854e:{UUID:c12ccfa5-8757-411e-9505-128190e9854e Name:11.4 Size:23465168896 Tags:[{Name:vanilla Size:18572955648} {Name:vanilla+port-forward-22 Size:18572955648} {Name:vanilla+port-forward-22+brew-git Size:21062053888} {Name:v1 Size:23465168896}]}]
```

You can see right near the end, after I push v1 onto the template, the size changes and an update is triggered for ONLY that template.

Reply:
See https://github.com/veertuinc/anka-prometheus-exporter/pull/20/commits/37995443f50b4770797c97687cd3cb808accf43d